### PR TITLE
feat: add update_user_info mutation

### PR DESCRIPTION
**Pull request by tdelabro**

### CODE REVIEW

File: api/src/domain/user_info/identity.rs [hunk 1]
```diff
@@ -0,0 +1,61 @@
+use derive_more::From;
+use juniper::{GraphQLEnum, GraphQLInputObject};
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Clone, Serialize, Deserialize, AsExpression, From, GraphQLInputObject)]
+pub struct IdentityInput {
```
Comment (AnthonyBuisset) [inline api/src/domain/user_info/identity.rs:6]: could this struct live in the presentation layer as it is only used there ?

File: api/src/domain/user_info/email.rs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+use derive_more::From;
+use email_address::EmailAddress;
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Clone, Serialize, Deserialize, AsExpression, From, FromToSql, FromSqlRow)]
+#[sql_type = "diesel::sql_types::Text"]
+#[serde(transparent)]
+pub struct Email(String);
+
+#[juniper::graphql_scalar(description = "A valid email address")]
```
Comment (AnthonyBuisset) [inline api/src/domain/user_info/email.rs:12]: I would rather have this implementation in the presentation layer, as it is how we decode it from graphql, not really a domain logic

File: api/src/domain/user_info/identity.rs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, AsExpression, FromToSql, FromSqlRow)]
+#[sql_type = "diesel::sql_types::Jsonb"]
+pub enum Identity {
+	Company(CompanyIdentity),
+	Person(PersonIdentity),
+}
+
+impl TryFrom<IdentityInput> for Identity {
```
Comment (AnthonyBuisset) [inline api/src/domain/user_info/identity.rs:19]: This blovk as well should be in the presentation layer as more related to `IdentityInput` than `Identity`

File: api/src/domain/user_info/mod.rs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+	location: Location,
+	email: Email,
+	payout_settings: PayoutSettings,
+}
+
+impl Entity for UserInfo {
+	type Id = UserId;
+}
+
+#[derive(
```
Comment (AnthonyBuisset) [inline api/src/domain/user_info/mod.rs:43]: could you please move this `Location` in a dedicated `location.rs` file like the identity ?

File: api/src/domain/user_info/payout_settings.rs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, AsExpression, FromToSql, FromSqlRow)]
+#[sql_type = "diesel::sql_types::Jsonb"]
+pub enum PayoutSettings {
+	WireTransfer(BankAddress),
+	EthTransfer(EthereumAddress),
+}
+
+impl TryFrom<PayoutSettingsInput> for PayoutSettings {
```
Comment (AnthonyBuisset) [inline api/src/domain/user_info/payout_settings.rs:25]: same remark as `IdentityInput`

File: api/src/routes/dto/user.rs [hunk 6]
```diff
@@ -0,0 +1,64 @@
+use crate::domain::{user, User as DomainUser};
```
Comment (AnthonyBuisset) [inline api/src/routes/dto/user.rs:1]: you might have some conflicts with Olivier's changes

File: domain/src/user/from_rocket_request.rs [hunk 7]
```diff
@@ -0,0 +1,49 @@
+use std::str::FromStr;
```
Comment (AnthonyBuisset) [inline domain/src/user/from_rocket_request.rs:1]: this looks like a dto, why can't it be in a presentation layer ?

**Review by AnthonyBuisset — COMMENTED** [review]
Could you please add some tests as well ?

File: api/src/graphql/mutation.rs [hunk 8]
```diff
@@ -123,4 +129,24 @@ impl Mutation {
...
+		email: Email,
+		payout_settings: PayoutSettingsInput,
+	) -> Result<Uuid> {
+		let user_id =
+			context.maybe_user_id.ok().map_err(|e| Error::NotAuthorized(e.to_string()))?;
+		let identity = Identity::try_from(identity).map_err(Error::InvalidRequest)?;
+		let payout_settings =
+			PayoutSettings::try_from(payout_settings).map_err(Error::InvalidRequest)?;
+
+		let user_info = UserInfo::new(user_id, identity, location, email, payout_settings);
```
Comment (Bernardstanislas) [inline api/src/graphql/mutation.rs:146]: This should be in a use case as we did for the other mutations

File: api/src/routes/dto/user.rs [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+	Admin,
+	IdentifiedUser {
+		lead_projects: HashSet<ProjectId>,
+		owned_budgets: HashSet<BudgetId>,
+	},
+	Public,
+}
+
+#[async_trait]
+impl<'r> FromRequest<'r> for User {
```
Comment (Bernardstanislas) [inline api/src/routes/dto/user.rs:19]: For the same reasons as email and identity, this must be unit tested

File: domain/src/user/from_rocket_request.rs [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+#[derive(Debug, Error)]
+pub enum IdFromRequestError {
+	#[error("{HASURA_USER_ID_HEADER_KEY} header is not present")]
+	Missing,
+	#[error("{HASURA_USER_ID_HEADER_KEY} header is invalid")]
+	Invalid,
+}
+
+pub struct MaybeId(Option<Id>);
```
Comment (Bernardstanislas) [inline domain/src/user/from_rocket_request.rs:24]: Seems fine to call it `OptionId` or `IdOption` directly, I would be confused to find in the codebase both `options` and `maybes` – which are the same things called differently by different schools of thought

File: domain/src/user/from_rocket_request.rs [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+		self.0
+	}
+
+	pub fn ok(&self) -> Result<Id, IdFromRequestError> {
+		self.0.ok_or(IdFromRequestError::Missing)
+	}
+}
+
+#[async_trait]
+impl<'r> FromRequest<'r> for MaybeId {
```
Comment (Bernardstanislas) [inline domain/src/user/from_rocket_request.rs:37]: This must be unit tested

File: macros/src/lib.rs [hunk 12]
```diff
@@ -17,9 +17,35 @@ pub fn diesel_repository(input: TokenStream) -> TokenStream {
 	diesel_repository::derive(input)
 }
 
-#[proc_macro_derive(NewtypeToSql, attributes(sql_type))]
+/// Parse a FromToSql derive macro.
```
Comment (Bernardstanislas) [inline macros/src/lib.rs:20]: ❤️

File: api/src/domain/user_info/identity.rs (hunk 1 above)
Comment (tdelabro) [reply api/src/domain/user_info/identity.rs:6]: done

File: api/src/domain/user_info/email.rs (hunk 2 above)
Comment (tdelabro) [reply api/src/domain/user_info/email.rs:12]: I cannot do this without, duplication of the struct.
Is that what we want?

File: api/src/domain/user_info/identity.rs (hunk 3 above)
Comment (tdelabro) [reply api/src/domain/user_info/identity.rs:19]: done

File: api/src/domain/user_info/mod.rs (hunk 4 above)
Comment (tdelabro) [reply api/src/domain/user_info/mod.rs:43]: done

File: api/src/domain/user_info/payout_settings.rs (hunk 5 above)
Comment (tdelabro) [reply api/src/domain/user_info/payout_settings.rs:25]: done

File: api/src/routes/dto/user.rs (hunk 6 above)
Comment (tdelabro) [reply api/src/routes/dto/user.rs:1]: look like no conflicts as for right now, may come when he merge?

File: domain/src/user/from_rocket_request.rs (hunk 7 above)
Comment (tdelabro) [reply domain/src/user/from_rocket_request.rs:1]: done

File: Makefile [hunk 13]
```diff
@@ -33,16 +36,16 @@ db/load-fixtures: db/up
 	@echo "It was generated on the `GIT_PAGER=cat git log -1 --format=%cd ./scripts/fixtures/latest.dump`"
 	@echo "To have a fresh dump, you can use db/update-staging-dump"
 
-migration/run: db/up
+db/migrate: db/up
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply Makefile:39]: right

File: Makefile [hunk 14]
```diff
@@ -33,16 +36,16 @@ db/load-fixtures: db/up
...
 	diesel migration run
 
-api/start: docker/up migration/run
+api/start: docker/up db/migrate
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply Makefile:42]: yup you are right

File: api/src/domain/user_info/email.rs [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+impl<S> GraphQLScalar for Email
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply api/src/domain/user_info/email.rs:13]: done for `from_input_value` which contains the core mechanic. Do you also want the two other methods to be tested ?

File: api/src/domain/user_info/payout_settings.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (tdelabro) [reply api/src/domain/user_info/payout_settings.rs:25]: can you point me to a snippet of code showing what you are referring to?

File: api/src/domain/user_info/payout_settings.rs [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(transparent)]
+pub struct EthereumAddress(String);
+
+#[juniper::graphql_scalar(
+	description = "A `0x` prefixed hexadecimal string representing 20 bytes of data"
+)]
+impl<S> GraphQLScalar for EthereumAddress
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply api/src/domain/user_info/payout_settings.rs:58]: done

File: api/src/graphql/mutation.rs (hunk 8 above)
Comment (tdelabro) [reply api/src/graphql/mutation.rs:146]: Can we not do it as we said this moring?
We will later on if the need arise

File: domain/src/user/from_rocket_request.rs (hunk 10 above)
Comment (tdelabro) [reply domain/src/user/from_rocket_request.rs:24]: No preferences on this. `OptionId` sounds perfectly good to me.

File: domain/src/user/from_rocket_request.rs (hunk 11 above)
Comment (tdelabro) [reply domain/src/user/from_rocket_request.rs:37]: done

File: macros/src/lib.rs (hunk 12 above)
Comment (tdelabro) [reply macros/src/lib.rs:20]: <3

File: api/src/domain/user_info/email.rs [hunk 17]
```diff
@@ -20,10 +18,39 @@ where
...
+mod test {
+	use super::Email;
+	use juniper::{DefaultScalarValue, FromInputValue, InputValue};
+	use rstest::rstest;
+
+	#[rstest]
+	#[case(InputValue::Scalar(DefaultScalarValue::String("abc42@gmail.com".to_string())), true)]
+	#[case(InputValue::Scalar(DefaultScalarValue::String("aBc@gMail.coM".to_string())), true)]
+	#[case(InputValue::Scalar(DefaultScalarValue::String("abc@gmail".to_string())), true)]
+	#[case(InputValue::Scalar(DefaultScalarValue::String("jsmith@[192.168.2.1]".to_string())), true)]
```
Comment (AnthonyBuisset) [inline api/src/domain/user_info/email.rs:42]: 😲

**Review by AnthonyBuisset — APPROVED** [review]

**Review by ofux — APPROVED** [review]

File: testing/cypress/e2e/simple_user.cy.js [hunk 18]
```diff
@@ -60,4 +60,60 @@ describe("As a simple user, I", () => {
...
         })
     });
+
+    it("can update my info", () => {
+        let email = "pierre.fabre@gmail.com";
+        let location = '{city: "Paris", country: "France", number: "4", postCode: "75008", street: "avenue des Champs Elysee"}';
+        let identity = '{type: PERSON, optPerson: {firstname: "Pierre", lastname: "Fabre"}}';
+        let payout_settings = '{type: ETHEREUM_ADDRESS, optEthAddress: "0x123"}';
+
+        let new_payout_settings = '{type: ETHEREUM_ADDRESS, optEthAddress: "0x456"}';
```
Comment (tdelabro) [inline testing/cypress/e2e/simple_user.cy.js:70]: I tried to use https://github.com/vkolgi/json-to-graphql-query instead but it's straight up not working. I opened an issue: https://github.com/vkolgi/json-to-graphql-query/issues/60
When fixed we could give it another try

File: testing/cypress.config.js [hunk 19]
```diff
@@ -6,6 +6,15 @@ const { defineConfig } = require("cypress");
 module.exports = defineConfig({
     e2e: {
         baseUrl: process.env.CYPRESS_BASE_URL,
+        setupNodeEvents(on, config) {
```
Comment (onlydust-contributor) [inline testing/cypress.config.js:9]: 😍

**Review by onlydust-contributor — APPROVED** [review]

**Pull request merged**